src/app/services/layout.service.ts

```diff
@@ -1,23 +1,320 @@
-    public layout(diagram: Run): void {
-        diagram.elements.forEach((el) => {
-            el.x =
-                Math.floor(Math.random() * LayoutService.RANGE) +
-                LayoutService.OFFSET;
-            el.y =
-                Math.floor(Math.random() * LayoutService.RANGE) +
-                LayoutService.OFFSET;
+    public layout(run: Run): void {
+        run.clearPositioningData();
+        //if run hast no cycles use sugiyama layout
+        if (!run.hasCycles()) {
+            const layers = this.assignLayers(run);
+            this.addBreakpoints(layers);
+            this.minimizeCrossing(layers);
+            this.calculatePosition(layers);
+        } else {
+            run.elements.forEach((el) => {
+                el.x =
+                    Math.floor(Math.random() * LayoutService.RANGE) +
+                    LayoutService.OFFSET;
+                el.y =
+                    Math.floor(Math.random() * LayoutService.RANGE) +
+                    LayoutService.OFFSET;
+            });
+        }
+    }
+
+    /**
+     * Sets the layer of all elements of the run
+     * All elements without incoming arcs are assigned to the next layer
+     * The outgoing arcs of all elements in the current layer are deleted to identify the next layer
+     * @param run run for which the layout is to be determined
+     * @returns layers with elements and breakpoints
+     */
+    private assignLayers(run: Run): Array<Element[]> {
+        const layers = new Array<Element[]>();
+        const elements = [...run.elements];
+        let arcs = [...run.arcs.filter((a) => a.sourceEl && a.targetEl)]; //use only arcs with valid source and target
+
+        while (elements.length > 0) {
+            const layer = new Array<Element>();
+            const elementsWithIncomingArcs = arcs.map((a) => a.targetEl);
+            //filter all elements without incoming arcs => add them to the current layer and remove their outgoing arcs
+            elements
+                .filter((e) => elementsWithIncomingArcs.indexOf(e) === -1)
+                .forEach((e) => {
+                    layer.push(e);
+                    elements.splice(elements.indexOf(e), 1);
+                    arcs = arcs.filter((a) => e.outogingArcs.indexOf(a) === -1);
+                });
+            layers.push(layer);
+        }
+        return layers;
+    }
+
+    /**
+     * Adds breakpoints to layers for arcs if there are layers between the source and target elements
+     * 1. Loop through all layers
+     *  2. Loop through all elements in the layer
+     *   3. Loop through all outgoing arcs of the element
+     *    4. check distance/layers between arc source and target
+     *     5. add breakpoint to arc for each enclosed layer
+     * @param layers layers with elements and breakpoints
+     */
+    private addBreakpoints(layers: Array<(Element | Breakpoint)[]>): void {
+        layers.forEach((layer, index) => {
```

Comment:
Warum benutzt du hier nicht einfach eine `for`-schleife? Du baust hier im Grunde selbst eine For-Schleife.